pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,291 @@
+            )
+            if is_independent:
+                self._basis.append(ps)
+        return self._basis
+
+    @staticmethod
+    def _add_if_independent(M, pauli_sentence, pw_to_idx, rank, num_pw):
+        r"""
+        Checks if ``pauli_sentence`` is linearly independent.
+
+        This is done in the following way: ``M`` (see description in class) is extended by ``pauli_sentence``.
+        If the added operator has a PauliWord (key) that is new to ``pw_to_idx``, then we have to add a new row
+        and already know that it has to be linearly independent.
+        If it contains the same PauliWords, we need to compute the new rank and compare it with the old rank.
+        If the rank is the same, the operator is linearly dependent and not added. Else, the rank is incrased by 1
+        and the extended M becomes our new M.
+
+        Args:
+            M (ndarray): coefficient matrix for current LIS
+            pauli_sentence (`~.PauliSentence`): Pauli sentence for which to add a column if independent
+            pw_to_idx (dict): map from ``PauliWord`` to row index in ``M``
+            rank (int): current rank of ``M``, equal to its number of columns
+            num_pw (int): current number of ``PauliWord``\ s, equal to the number of rows in ``M``
+        Returns:
+            ndarray: updated coefficient matrix for the LIS
+            dict: updated map from ``PauliWord`` to row index in ``M``. Includes new ``PauliWord`` keys
+                from the input ``pauli_sentence`` if it was linearly independent
+            int: updated rank/number of columns of ``M``
+            int: updated number of ``PauliWord``\ s/number of rows of ``M``
+            bool: whether ``pauli_sentence`` was linearly independent and its column was added to ``M``
+        """
+        new_pws = [pw for pw in pauli_sentence.keys() if pw not in pw_to_idx]
+        new_num_pw = num_pw + len(new_pws)
+
+        if new_num_pw < rank + 1:
+            # Can't span rank+1 independent vectors in fewer than rank+1 dimensions
+            # The input PauliSentence must have been linearly dependent
+            return M, pw_to_idx, rank, num_pw, False
+
+        M = np.pad(M, ((0, new_num_pw - num_pw), (0, 1)))
+
+        # If there are new PauliWords (i.e. new basis vectors), the candidate vector must be linearly independent
+        if new_num_pw > num_pw:
+            new_pw_to_idx = copy(pw_to_idx)
+            for i, pw in enumerate(new_pws, start=num_pw):
+                new_pw_to_idx[pw] = i
+            # Add new PauliSentence entries to matrix
+            for pw, value in pauli_sentence.items():
+                M[new_pw_to_idx[pw], rank] = value
+
+            return M, new_pw_to_idx, rank + 1, new_num_pw, True
+
+        # Add new PauliSentence entries to matrix
+        for pw, value in pauli_sentence.items():
+            M[pw_to_idx[pw], rank] = value
+
+        # Check if new vector is proportional to any of the previous vectors
```

Comment:
I think if we maintained an orthonormal `M`, we could check if a new vector `v` is spanned by `M` as follows
```suggestion
    v /= np.linalg.norm(v)
    v = v - Q@(np.conj(Q).T @ v)
    if np.linalg.norm(v) > tol:
```
Otherwise, we can also build an orthonormal `M` on-the-fly as follows
```suggestion
    v /= np.linalg.norm(v)
    v = v - A@np.linalg.inv(np.conj(A.T) @ A)@np.conj(A).T @ v
    if np.linalg.norm(v) > tol:
```

Reply:
You reckon the second option is faster than `_is_any_col_propto_last(M)`?

Reply:
Some ballpark numbers. 

For small M (n=2)
projection: 11.7 µs ± 122 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
prop check: 18.7 µs ± 17.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

large M (N=30)
projection: 271 µs ± 65.3 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
prop check: 30.3 µs ± 73.6 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

```
n=30
v1 = VSpace([
    PauliSentence({PauliWord({i:"X", i+1:"X"}): 1., PauliWord({i:"Y", i+1:"Y"}): 1.}) for i in range(30)
])
M = v1._M

%%timeit
v = M[:, -1]
v /= np.linalg.norm(v)
v = v - M@np.linalg.inv(np.conj(M.T) @ M)@np.conj(M).T @ v

%timeit _is_any_col_propto_last(M)
```

Reply:
`projection` is meant to replace the whole linear dependence check (so `matrix_rank`, etc.), not just `_is_any_col_propto_last`.

Reply:
If `if np.linalg.norm(v) > tol:` (say with `tol = 1e-15`) then `v` is not linearly dependent and added to the basis. Otherwise throw away and return.